src/Compilers/CSharp/Test/Semantic/Semantics/SimpleProgramsTests.cs

```diff
@@ -5094,5 +5105,874 @@ void verifyModel(bool ignoreAccessibility, string expectedType)
+            {
+                context.RegisterOperationBlockAction(Handle);
+            }
+
+            private void Handle(OperationBlockAnalysisContext context)
+            {
+                Assert.Equal("<simple-program-entry-point>", context.OwningSymbol.ToTestDisplayString());
+                Assert.Equal(SyntaxKind.CompilationUnit, context.OperationBlocks.Single().Syntax.Kind());
+
+                switch (context.OperationBlocks.Single().Syntax.ToString())
+                {
+                    case "System.Console.WriteLine(1);":
+                        Interlocked.Increment(ref FireCount1);
+                        break;
+                    case "System.Console.WriteLine(2);":
+                        Interlocked.Increment(ref FireCount2);
+                        break;
+                    default:
+                        Assert.True(false);
+                        break;
+                }
+            }
+        }
+
+        [Fact]
+        public void AnalyzerActions_06()
+        {
+            var text1 = @"System.Console.WriteLine(1);";
+
+            var analyzer = new AnalyzerActions_06_Analyzer();
+            var comp = CreateCompilation(text1, options: TestOptions.DebugExe, parseOptions: DefaultParseOptions);
+            comp.GetAnalyzerDiagnostics(new[] { analyzer }, null).Verify();
+
+            Assert.Equal(1, analyzer.FireCount1);
+            Assert.Equal(0, analyzer.FireCount2);
+
+            var text2 = @"System.Console.WriteLine(2);";
+
+            analyzer = new AnalyzerActions_06_Analyzer();
+            comp = CreateCompilation(new[] { text1, text2 }, options: TestOptions.DebugExe, parseOptions: DefaultParseOptions);
+            comp.GetAnalyzerDiagnostics(new[] { analyzer }, null).Verify();
+
+            Assert.Equal(1, analyzer.FireCount1);
+            Assert.Equal(1, analyzer.FireCount2);
+        }
+
+        private class AnalyzerActions_06_Analyzer : DiagnosticAnalyzer
+        {
+            public int FireCount1;
+            public int FireCount2;
+
+            private static readonly DiagnosticDescriptor Descriptor =
+               new DiagnosticDescriptor("XY0000", "Test", "Test", "Test", DiagnosticSeverity.Warning, true, "Test", "Test");
+
+            public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+            => ImmutableArray.Create(Descriptor);
+
+            public override void Initialize(AnalysisContext context)
+            {
+                context.RegisterOperationBlockStartAction(Handle);
+            }
+
+            private void Handle(OperationBlockStartAnalysisContext context)
+            {
+                Assert.Equal("<simple-program-entry-point>", context.OwningSymbol.ToTestDisplayString());
+                Assert.Equal(SyntaxKind.CompilationUnit, context.OperationBlocks.Single().Syntax.Kind());
+
+                switch (context.OperationBlocks.Single().Syntax.ToString())
```

Comment:
For all the RegisterXXXStartAction, consider registering nested actions on the start context, such as `context.RegisterOperationAction` here and then move this switch statement into that callback.

Reply:
> For all the RegisterXXXStartAction, consider registering nested actions on the start context, such as context.RegisterOperationAction here and then move this switch statement into that callback.

Added AnalyzerActions_12, AnalyzerActions_13

---
In reply to: [403211428](https://github.com/dotnet/roslyn/pull/43038#discussion_r403211428) [](ancestors = 403211428)